File: Year 2/Semester 4/Programming and Projection Environments/Week 9/ProtobuffProjectJava/ServerRPC/src/main/java/gRPCProject/server/ServicesImpl.java
package gRPCProject.server;

import gRPCProject.domain.*;
import gRPCProject.persistance.repository.IArtistRepository;
import gRPCProject.persistance.repository.IEmployeeRepository;
import gRPCProject.persistance.repository.IShowRepository;
import gRPCProject.persistance.repository.ITicketRepository;
import gRPCProject.service.IObserver;
import gRPCProject.service.IService;
import gRPCProject.service.MyException;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;


public class ServicesImpl implements IService {

    private final IArtistRepository<Integer, Artist> artistRepository;
    private final IEmployeeRepository<Integer, Employee> employeeRepository;
    private final IShowRepository<Integer, Show> showRepository;
    private final ITicketRepository<Integer, Ticket> ticketRepository;
    private final Map<Integer, IObserver> loggedEmployees;

    public ServicesImpl(IArtistRepository<Integer, Artist> artistRepository,
                        IEmployeeRepository<Integer, Employee> employeeRepository,
                        IShowRepository<Integer, Show> showRepository,
                        ITicketRepository<Integer, Ticket> ticketRepository) {
        this.artistRepository = artistRepository;
        this.employeeRepository = employeeRepository;
        this.showRepository = showRepository;
        this.ticketRepository = ticketRepository;
        loggedEmployees = new ConcurrentHashMap<>();
    }


    @Override
    public synchronized void addTickets(Ticket ticket) throws MyException {
        ticketRepository.addTicket(ticket);
        for (IObserver employeeO: loggedEmployees.values()) {
            employeeO.ticketAdded(ticket);
        }
    }

    @Override
    public synchronized Employee authenticateEmployee(Employee employee, IObserver employeeObserver) throws MyException {
        Employee authenticateEmployee = employeeRepository.authenticateEmployee(employee.getUsername(), employee.getPassword());
        if (authenticateEmployee!=null){
            if(loggedEmployees.get(authenticateEmployee.getId())!=null)
                throw new MyException("User already logged in.");
            loggedEmployees.put(authenticateEmployee.getId(), employeeObserver);
        }else
            throw new MyException("Authentication failed.");
        return authenticateEmployee;
    }

    @Override
    public synchronized void logout(Employee employee, IObserver employeeObserver) throws MyException {
        IObserver localClient= loggedEmployees.remove(employee.getId());
        if (localClient==null)
            throw new MyException("User "+employee.getId()+" is not logged in.");
    }

    @Override
    public synchronized Iterable<ShowDTO> getAllShowsDTO() {
        return showRepository.getAllShowsDTO();
    }

    @Override
    public synchronized Iterable<Artist> getAllArtists() throws MyException {
        return artistRepository.getAll();
    }

    @Override
    public synchronized Artist findArtist(Integer id) throws MyException {
        return artistRepository.findById(id);
    }
}

File: Year 2/Semester 4/Programming and Projection Environments/Week 9/ProtobuffProjectCSharp/SeventhWeekCSharpServerClient/ServicesImpl.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using ProtobuffProject.domain;
using ProtobuffProject.repository.interfaces;
using ProtobuffProject.services;

namespace ProtobuffProject.server
{
    public class ServicesImpl : IServices
    {
        private readonly IArtistRepository artistRepository;
        private readonly IEmployeeRepository employeeRepository;
        private readonly IDictionary<int, IObserver> loggedEmployees;
        private readonly IShowRepository showRepository;
        private readonly ITicketRepository ticketRepository;

        public ServicesImpl(IArtistRepository artistRepository, IEmployeeRepository employeeRepository,
            IShowRepository showRepository, ITicketRepository ticketRepository)
        {
            this.artistRepository = artistRepository;
            this.employeeRepository = employeeRepository;
            this.showRepository = showRepository;
            this.ticketRepository = ticketRepository;
            loggedEmployees = new Dictionary<int, IObserver>();
        }

        public void AddTickets(Ticket<int> ticket)
        {
            ticketRepository.AddTicket(ticket);
            foreach (var employeeO in loggedEmployees.Values) Task.Run(() => employeeO.TicketAdded(ticket));
        }

        public Employee<int> AuthenticateEmployee(Employee<int> employee, IObserver employeeObserver)
        {
            var authenticateEmployee = employeeRepository.AuthenticateEmployee(employee.Username, employee.Password);
            if (authenticateEmployee != null)
            {
                if (loggedEmployees.ContainsKey(authenticateEmployee.Id))
                    throw new MyException("User already logged in.");
                loggedEmployees[authenticateEmployee.Id] = employeeObserver;
            }
            else
            {
                throw new MyException("Authentication failed.");
            }

            return authenticateEmployee;
        }

        public void Logout(Employee<int> employee)
        {
            var localClient = loggedEmployees[employee.Id];
            if (localClient == null)
                throw new MyException("User " + employee.Id + " is not logged in.");
            loggedEmployees.Remove(employee.Id);
        }

        public List<ShowDTO<int>> GetAllShowsDTO()
        {
            return showRepository.GetAllShows();
        }

        public List<Artist<int>> GetAllArtists()
        {
            return artistRepository.GetAll();
        }

        public Artist<int> FindArtist(int id)
        {
            return artistRepository.FindById(id);
        }
    }
}